src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala

```diff
@@ -20,31 +20,31 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
 
   override def compute(split: Partition, context: TaskContext): Iterator[Row] = {
     val driver = config.driver()
-    val session = driver.session()
+    val session = driver.session(config.sessionConfig())
 
     try {
-      val result: StatementResult = session.run(query, parameters.toMap.mapValues(_.asInstanceOf[AnyRef]).asJava)
-
-      result.asScala.map(record => {
-        val keyCount = record.size()
-
-        val res = if (keyCount == 0) Row.empty
-        else if (keyCount == 1) Row(convert(record.get(0).asObject()))
-        else {
-          val builder = Seq.newBuilder[AnyRef]
-          var i = 0
-          while (i < keyCount) {
-            builder += convert(record.get(i).asObject())
-            i = i + 1
-          }
-          Row.fromSeq(builder.result())
-        }
-        if (!result.hasNext) {
-          if (session.isOpen) session.close()
-          driver.close()
-        }
-        res
-      })
+      val runner = new TransactionWork[Iterator[Row]]() {
+        override def execute(tx: Transaction): Iterator[Row] = {
+          val result: Result = tx.run(query, parameters.toMap.mapValues(_.asInstanceOf[AnyRef]).asJava)
+          result.list().asScala.map(record => {
+            val keyCount = record.size()
+
+            val res = if (keyCount == 0) Row.empty
+            else if (keyCount == 1) Row(convert(record.get(0).asObject()))
+            else {
+              val builder = Seq.newBuilder[AnyRef]
+              var i = 0
+              while (i < keyCount) {
+                builder += convert(record.get(i).asObject())
+                i = i + 1
+              }
+              Row.fromSeq(builder.result())
+            }
+            res
+          })
+        }.iterator
+      }
+      session.readTransaction(runner)
     } finally {
```

Comment:
I think this needs experimentation.   I could be wrong about this so we don't necessarily need a change but when I was playing with my own private branch adapting spark for 4.0, these "finally close the session" blocks were a big problem.   This needs more testing.

Reason is that under the new reactive driver setup, the semantics of how/when records are consumed has changed.   Spark itself is so lazy, when you exit this block the very first record hasn't been pulled yet, but you're closing the session.   In my experimentation this broke badly because basically you'd always fail to get data back from neo4j saying that the driver or the session was closed.

For unit tests on this you might need some delay, and record sets with > 1 coming back.   Or you can create a test build, play with it in zeppelin

I'll not put the same comment on every finally block, but this pattern (run the query, return, close the session/driver in a finally block) is common in the code base.

Reply:
See "important note on semantics" in the new java driver API docs -- I think this interacting with spark's laziness is specifically problematic.  https://neo4j.com/docs/api/java-driver/4.0/org/neo4j/driver/QueryRunner.html

Reply:
@moxious thanks a lot for the feedback. We definitely need to improve this, I have in mind to do it with a new release that supports Spark Datasource APIs.

About transaction management, looking at the code I invoke `Result#list` which store the whole result for later use
https://neo4j.com/docs/api/java-driver/4.0/org/neo4j/driver/Result.html#list--

This is the only way that I found to interact with the db that allows us to use transaction functions (that should allow us to work properly with Aura)

Wdyt?

Reply:
it's definitely a tricky area. I think we should discuss with the drivers team what's the best interaction pattern that gives us all

- lazyness / reactive
- transaction functions
- ensure closing driver / session to avoid leaks
- support sparks lazy evaluation
- avoid pre-materialization of results

Reply:
can we organize a chat with them?

Reply:
I'll ask if we can put together that chat.  I don't have an opinion on a solution approach because this stuff is too new to me and I don't understand the consequences either way.